### PR TITLE
Update WFC and add extractOptimizableObjectRefs

### PR DESCRIPTION
**Pull request by ye-luo**
## Proposed changes
Add extractOptimizableObjectRefs()
~~Update WFC. a) add getClassName() and isFermion() as virtual.~~ separate PR #4170

## What type(s) of changes does this code introduce?
- New feature

### Does this introduce a breaking change?
- No

## What systems has this change been tested on?
epyc-server

## Checklist
- Yes. This PR is up to date with current the current state of 'develop'

### CODE REVIEW

File: src/QMCWaveFunctions/TrialWaveFunction.h [hunk 1]
```diff
@@ -144,8 +144,13 @@ class TrialWaveFunction
   // Finally, the call to resetParameters progates the new values (step 3).
   // The call to checkOutVariables is a prerequisite for resetParameters to set the local values successfully.
 
+  /** extract underlying OptimizableObject references
+   * @param opt_obj_refs aggregated list of optimizable object references
+   */
+  void extractOptimizableObjectRefs(UniqueOptObjRefs& opt_obj_refs);
```
Comment (markdewing) [inline src/QMCWaveFunctions/TrialWaveFunction.h:150]: "Refs" probably shouldn't be in the name of the function - that puts too much implementation detail in the name.

Comment (ye-luo) [reply src/QMCWaveFunctions/TrialWaveFunction.h:150]: Good suggestions. However, I would defer the change until the PRs in my pockets gets merged and then make the change once.

**Comment by markdewing** [comment]
Could you move all the small changes (getClassName, adding accessor functions, formatting changes) to a separate PR?

**Comment by ye-luo** [comment]
@markdewing I took out a few commits in #4170

File: src/QMCWaveFunctions/TrialWaveFunction.h (hunk 1 above)
Comment (ye-luo) [reply src/QMCWaveFunctions/TrialWaveFunction.h:150]: I have changed the API signature in a later PR, opt_obj_refs is a return value directly.

**Review by PDoakORNL — APPROVED** [review]
This looks good to me, it will shock no one that major state changes occur in the put method and that an object using that legacy API will be invalid until put is called.

File: src/QMCDrivers/WFOpt/QMCCostFunctionBase.cpp [hunk 2]
```diff
@@ -399,6 +396,17 @@ bool QMCCostFunctionBase::put(xmlNodePtr q)
     }
     cur = cur->next;
   }
+
+  /// survey all the optimizable objects
```
[the remark this answers is not in the document]
Comment (ye-luo) [reply src/QMCDrivers/WFOpt/QMCCostFunctionBase.cpp:400]: This is survey. I only printout what is found by extractOptimizableObjectRefs.
checkInVariables on line 412 is still the old way. In #4168, checkInVariablesExclusive will be called from opt_obj_refs.

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**